packages/compiler/src/core/name-resolver.ts

```diff
@@ -0,0 +1,1331 @@
+ * failed and move on. Even in cases where we could give a good error we chose
+ * not to in order to uniformly handle error reporting in the checker.
+ *
+ * Name resolution has three sub-phases:
+ *
+ * 1. Merge namespace symbols and decorator implementation/declaration symbols
+ * 2. Resolve using references to namespaces and create namespace-local bindings
+ *    for used symbols
+ * 3. Resolve type references and bind members
+ *
+ * The reference resolution and member binding phase implements a deferred
+ * resolution strategy. Often we cannot resolve a reference without binding
+ * members, but we often cannot bind members without resolving references. In
+ * such situations, we stop resolving or binding the current reference or type
+ * and attempt to resolve or bind the reference or type it depends on. Once we
+ * have done so, we return to the original reference or type and complete our
+ * work.
+ *
+ * This is accomplished by doing a depth-first traversal of the reference graph.
+ * On the way down, we discover any dependencies that need to be resolved or
+ * bound for the current node, and recurse into the AST nodes, so that on the
+ * way back up, all of our dependencies are bound and resolved and we can
+ * complete. So while we start with a depth-first traversal of the ASTs in order
+ * to discover work to do, most of the actual work is done while following the
+ * reference graph, binding and resolving along the way. Circular references are
+ * discovered during the reference graph walk and marked as such. Symbol and
+ * node links are used to ensure we never resolve the same reference twice. The
+ * checker implements a very similar algorithm to evaluate the types of the
+ * program.
+ **/
+
+import { isArray, Mutable, mutate } from "../utils/misc.js";
+import { createSymbol, createSymbolTable } from "./binder.js";
+import { compilerAssert } from "./diagnostics.js";
+import { inspectSymbol } from "./inspector.js";
+import { createDiagnostic } from "./messages.js";
+import { visitChildren } from "./parser.js";
+import { Program } from "./program.js";
+import { getFullyQualifiedSymbolName } from "./type-utils.js";
+import {
+  AliasStatementNode,
+  AugmentDecoratorStatementNode,
+  DecoratorExpressionNode,
+  EnumStatementNode,
+  Expression,
+  IdentifierNode,
+  InterfaceStatementNode,
+  IntersectionExpressionNode,
+  MemberExpressionNode,
+  ModelExpressionNode,
+  ModelPropertyNode,
+  ModelStatementNode,
+  NamespaceStatementNode,
+  Node,
+  NodeFlags,
+  NodeLinks,
+  OperationStatementNode,
+  ProjectionDecoratorReferenceExpressionNode,
+  ProjectionStatementNode,
+  ResolutionResult,
+  ResolutionResultFlags,
+  ScalarStatementNode,
+  Sym,
+  SymbolFlags,
+  SymbolLinks,
+  SymbolTable,
+  SyntaxKind,
+  TemplateParameterDeclarationNode,
+  TypeReferenceNode,
+  TypeSpecScriptNode,
+  UnionStatementNode,
+} from "./types.js";
+
+export interface NameResolver {
```

Comment:
Alternative name which could be more accurate `Linker`

Reply:
No strong opinion from me. Follow your heart!